aten/src/ATen/native/RNN.cpp

```diff
@@ -324,55 +349,97 @@ struct LayerOutput {
 template<typename io_type, typename hidden_type, typename param_type>
 struct Layer {
   using output_type = LayerOutput<io_type, hidden_type>;
-  virtual ~Layer() {} // This is really dumb, but enables projects with -Wnon-virtual-dtor to compile...
-  virtual output_type operator()(const io_type& input, const hidden_type& input_hidden, const param_type& params) const = 0;
+
+  virtual ~Layer() {} // This is really dumb, but enables projects with
+                      // -Wnon-virtual-dtor to compile...
+  virtual output_type operator()(
+      const io_type& input,
+      const hidden_type& input_hidden,
+      const param_type& params) const = 0;
 };
 
 template<typename hidden_type, typename cell_params>
 struct FullLayer : Layer<Tensor, hidden_type, cell_params> {
-  using output_type = typename Layer<Tensor, hidden_type, cell_params>::output_type;
+  using output_type =
+      typename Layer<Tensor, hidden_type, cell_params>::output_type;
   using unstacked_output_type = LayerOutput<std::vector<Tensor>, hidden_type>;
 
   FullLayer(Cell<hidden_type, cell_params>& cell)
     : cell_(cell) {};
 
-  unstacked_output_type operator()(std::vector<Tensor> step_inputs, const hidden_type& input_hidden, const cell_params& params) const {
+  unstacked_output_type operator()(
+      const std::vector<Tensor>& step_inputs,
+      const hidden_type& input_hidden,
+      const cell_params& params,
+      bool pre_compute_input = false) const {
     std::vector<Tensor> step_outputs;
     auto hidden = input_hidden;
-    for (size_t i = 0; i < step_inputs.size(); i++) {
-      hidden = cell_(step_inputs[i], hidden, params);
-      step_outputs.push_back(hidden_as_output(hidden));
+    for (const auto& input : step_inputs) {
+      hidden = cell_(input, hidden, params, pre_compute_input);
+      step_outputs.emplace_back(hidden_as_output(hidden));
     }
     return {step_outputs, hidden};
   }
 
-  output_type operator()(const Tensor& inputs, const hidden_type& input_hidden, const cell_params& params) const override {
+  output_type operator()(
+      const Tensor& inputs,
+      const hidden_type& input_hidden,
+      const cell_params& params) const override {
+    if (inputs.device().is_cpu()) {
```

Comment:
What is wrong with the same approach in case of the GPU?

Reply:
I will apply it in the next PR for GPU. Originally on GPU we do matmul instead of linear for input first, and fuse the bias part in at::_thnn_fused_lstm_cell function. So I think it will be better to make change on CPU only first then apply the GPU change in case the PR to be too large.